libnmstate/nm/applier.py

```diff
@@ -85,23 +86,73 @@ def prepare_edited_ifaces_configuration(ifaces_desired_state):
     return con_profiles
 
 
-def set_ifaces_admin_state(ifaces_desired_state):
+def set_ifaces_admin_state(ifaces_desired_state, con_profiles=()):
+    """
+    Control interface admin state by activating, deactivating and deleting
+    devices connection profiles.
+
+    The `down` and `absent` states are handled in the same manner (temporary)
+    by deactivating the device and deleting the connection profile.
+
+    For new virtual devices, the `up` state is handled by activating the
+    new connection profile. For existing devices, the device is activated,
+    leaving it to choose the correct profile.
+
+    In order to activate correctly the interfaces, the order is significant:
+    - New interfaces (virtual interfaces).
+    - Master interfaces.
+    - All the rest.
+    """
+    new_ifaces = _get_new_ifaces(con_profiles)
+    new_ifaces_to_activate = set()
+    master_ifaces_to_activate = set()
     devs_actions = {}
+
     for iface_desired_state in ifaces_desired_state:
-        devs = _get_affected_devices(iface_desired_state)
-        if iface_desired_state['state'] == 'up':
-            devs_actions.update({dev: (device.activate,) for dev in devs})
-        elif iface_desired_state['state'] in ('down', 'absent'):
-            devs_actions.update(
-                {dev: (device.deactivate, device.delete) for dev in devs})
+        nmdev = device.get_device_by_name(iface_desired_state['name'])
+        if not nmdev:
+            ifname = iface_desired_state['name']
+            if ifname in new_ifaces and iface_desired_state['state'] == 'up':
+                    new_ifaces_to_activate.add(ifname)
         else:
-            raise UnsupportedIfaceStateError(iface_desired_state)
+            if iface_desired_state['state'] == 'up':
+                master_iface_types = ovs.BRIDGE_TYPE, bond.BOND_TYPE
+                if iface_desired_state['type'] in master_iface_types:
+                    master_ifaces_to_activate.add(nmdev)
+                else:
+                    devs_actions[nmdev] = (device.activate,)
+            elif iface_desired_state['state'] in ('down', 'absent'):
+                nmdevs = _get_affected_devices(iface_desired_state)
+                for nmdev in nmdevs:
+                    devs_actions[nmdev] = (device.deactivate, device.delete)
+            else:
+                raise UnsupportedIfaceStateError(iface_desired_state)
+
+    for ifname in new_ifaces_to_activate:
+        device.activate(
+            dev=None,
+            fetch_remote_connection_func=(
+                lambda: nmclient.client().get_connection_by_id(ifname))
```

Comment:
It seems that activate() could be easily be simplified by just accepting a connection id instead of the fetch function.

Reply:
The function is triggered only later by the mainloop, there is no connection ID at this time, it has not been created yet (the connection is created in the mainloop as well).

Reply:
> his might true for now, but there is no document of NM confirmed this.

It is document in our code:
https://github.com/nmstate/nmstate/blob/master/libnmstate/nm/connection.py#L105

> How about use nmclient.activate_connection_async where you can define device without connection/profile ID?

I think nmstate is supposed to be specific when controlling the interfaces. I am not clear how we could do that without a unique identifier (we have today the device and the profile ID).
But there may be a need to add other identifiers in the future, if and when that  happens we could add it like we added this one.

Reply:
Ahh, you are absolutely right. I was blind.
Fixed! Thank you!